Re-implemented the ecvt function.

mazong1123 opened the pull request:
Instead of leveraging snprintf, re-implement the ecvt function according to the paper: https://www.cs.indiana.edu/~dyb/pubs/FP-Printing-PLDI96.pdf

Note:
1. This commit won't fix any existing bug.
2. This is a raw implementation of the paper. The performance on Linux only gain 10%. We could tune the performance further.

**UPDATE**:

Now the new implementation is at least 2x faster than the old one.

Fix #10651

mazong1123 commented:
@tarekgh @jkotas This commit is a raw implementation of the paper https://www.cs.indiana.edu/~dyb/pubs/FP-Printing-PLDI96.pdf . The performance not gained so much (only 10%). I'll tune the performance further.

I want to go through the CI checks so that if there's any issue I can fix them earlier.

Note I did not change the "15 digit first, then 17 digit precision" round trip conversion rule. I think that should be done in another PR.

mazong1123 commented:
I ran some tests on this PR. Following are the details:

## Accuracy

I ran following code against this PR:
```csharp
using System;

namespace dtost
{
    class Program
    {
        static void TestDoubleConversion(double number)
        {
            string str = number.ToString("R");
            double back = double.Parse(str);
            if (number != back)
            {
                Console.WriteLine("Test failed: converted str: {0}. expected double {1}", back, number);
            }
            else
            {
                Console.WriteLine("Pass");
            }
        }

        static void Main(string[] args)
        {
            TestDoubleConversion(104234.343);
            TestDoubleConversion(-1.7976931348623157E+308);
            TestDoubleConversion(1.7976931348623157e+308);
            TestDoubleConversion(1.7976931348623157e+308);
            TestDoubleConversion(70.9228162514264339123);
            TestDoubleConversion(3.1415926535897937784612345);
            TestDoubleConversion(29999999999999792458.0);
            TestDoubleConversion(1000.4999999999999999999);
            TestDoubleConversion(1000.9999999999999999999);
            TestDoubleConversion(999.99999999999999999999);
            TestDoubleConversion(0.84551240822557006);
            TestDoubleConversion(1.0);
            TestDoubleConversion(10.0);
            TestDoubleConversion(0);
            TestDoubleConversion(0.0);
            TestDoubleConversion(0.0000000000000199);
            TestDoubleConversion(-0.0000123);
        }
    }
}
```

The output is:

```
Pass
Pass
Pass
Pass
Pass
Pass
Pass
Pass
Pass
Pass
Test failed: converted str: 0.84551240822557. expected double 0.84551240822557
Pass
Pass
Pass
Pass
Pass
Pass
```
Ran the code under current .NET Core release got exactly the same result. The only failure is caused by a known issue https://stackoverflow.com/questions/24299692/why-is-a-round-trip-conversion-via-a-string-not-safe-for-a-double 

So the accuracy is the same as current .NET Core release.

## Performance

I ran following micro-benchmark:
```csharp
using System;
using System.Diagnostics;
using System.Globalization;

namespace hlold
{
    class Program
    {
        static void Main(string[] args)
        {
            //double number = 104234.343;
            double number = double.MinValue / 2;
            CultureInfo cultureInfo = new CultureInfo("fr");
            for (int i = 0; i < 20; i++)
            {
                Stopwatch watch = new Stopwatch();
                watch.Start();
                for (int j = 0; j < 10000; j++)
                {
                    number.ToString(cultureInfo); number.ToString(cultureInfo); number.ToString(cultureInfo);
                    number.ToString(cultureInfo); number.ToString(cultureInfo); number.ToString(cultureInfo);
                    number.ToString(cultureInfo); number.ToString(cultureInfo); number.ToString(cultureInfo);
                }

                Console.Write(watch.ElapsedMilliseconds);
                Console.WriteLine();
            }
        }
    }
}

```
### Output of new implementation:

```
191
189
188
190
192
190
188
189
187
190
192
187
188
187
188
189
192
188
187
187
```

### Output of old implementation:

```
392
382
381
383
386
400
390
381
381
387
390
382
382
393
386
382
386
386
390
382
```

So the new implementation is at least 2x faster than the old one. I think we have proved that the algorithm is much more efficient than the old one.

tarekgh commented:
@dotnet-bot help

tarekgh commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline

tarekgh commented:
@mazong1123 this is fantastic. the results look promising. I have enabled the whole corefx repo to run against your changes to know if we get any regression case. Also I am not seeing we used a big static data which is a good news too. if corefx goes fine we can move to the next step to check with the legal about using the algorithm and do official code review then we can finalize the work. 

Thanks a lot for getting this done. good work!

tarekgh commented:
@safern could you please take a look at the CI falure in ComponentModel? 

Ubuntu x64 Checked Build and Test (Jit - CoreFx)

https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_ubuntu_corefx_baseline_prtest/62/consoleFull#-21281798079fe3b83-f408-404c-b9e7-9207d232e5fc

this is not urgent but it'll be good if you have a look when you have a chance.


@mazong1123 you may ignore the failure in "Ubuntu x64 Checked Build and Test (Jit - CoreFx)" for now as it looks unrelated to your changes.

safern commented:
> @safern could you please take a look at the CI falure in ComponentModel?

Will do later :)

tannergooding commented:
I'm going to give the implementation a review pass sometime later this evening.

tarekgh commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+    const UINT32* pRhsEnd = pRhsCurrent + length;
+
+    while (pRhsCurrent != pRhsEnd)
+    {
```
can we just memcpy this part instead of looping?

mikedn replied:
It may be best to measure to see if there's any performance difference.

tannergooding replied:
`memcpy` should do the right thing, regardless of size. That being said, the current max size is only `35` bytes so the perf difference is barely measurable. But, this size is way too small and should be well over 138 bytes (with exact size varying based on the maximum number of digits we consider).

tarekgh replied:
why we don't just memcpy considering m_len even if you think the perf is barely measurable? 

---
In reply to: [128531411](https://github.com/dotnet/coreclr/pull/12894#discussion_r128531411) [](ancestors = 128531411)

tarekgh commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+
+    UINT32 lhsValue = lhs.m_blocks[0];
+
+    if (lhsValue > value || lhs.m_len > 1)
```
>lhs.m_len > 1 [](start = 28, length = 13)

nit: moving this condition (lhs.m_len > 1) before the previous line would be better

tarekgh commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+
+    while (true)
+    {
+        if (*pCurrentLeftBlock > *pCurrentRightBlock)
```
nit: maybe we can do it as 

INT64 diff = (INT64) (*pCurrentLeftBlock) - (INT64) (*pCurrentRightBlock);
if (diff != 0)
   return diff;

tannergooding replied:
:+1:, but we should validate that this doesn't perform any more poorly on 32-bit platforms

tarekgh commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+    const UINT32* pCurrentLeftBlock = lhs.m_blocks + lastIndex;
+    const UINT32* pCurrentRightBlock = rhs.m_blocks + lastIndex;
+
+    while (true)
```
>true) [](start = 11, length = 5)

nit: while (pCurrentLeftBlock <= pLeftStartBlock) is better here and remove if (pCurrentLeftBlock == pLeftStartBlock) from the loop

mazong1123 replied:
I think you meant `while (pCurrentLeftBlock >= pLeftStartBlock)` :). I'll take this optimization.

tarekgh commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+
+void BigNum::ShiftLeft(UINT64 input, int shift, BigNum& output)
+{
+    int shiftBlocks = shift / 32;
```
>int shiftBlocks = shift / 32; [](start = 4, length = 29)

shift << 5

tannergooding replied:
👍, although the compiler should also be doing this optimization for us.

tannergooding replied:
Actually, we should probably just leave this. The compiler will optimize the next statement as well (into either a single `idiv` or into a couple of shifts and a subtract)

mikedn replied:
Using `/` and `%` is fine, no native compiler will generate actual div instructions for this code. However, I would suggest making `shift` unsigned because unsigned division by constant is cheaper than signed division. It may be the native compiler will figure this one too but it requires interprocedural optimizations so it's somewhat less certain.

mazong1123 replied:
@mikedn Good catch. I think make `shift` unsigned should be more reasonable here.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+        realExponent = -1074;
+        mantissaHighBitIdx = BigNum::LogBase2(realMantissa);
+    }
+
```
You can shortcut `Zero`, `Infinity`, `Indeterminate`, and `NaN` by this point

mazong1123 replied:
These special cases has already been handled before passing to the method: https://github.com/dotnet/coreclr/pull/12894/files/b617b7928f47263980752318951663b7faa79dee#diff-0d0ac521b3af6ae20f3148fa1bee21f1L146

I think adding asserts is worth though.

tannergooding replied:
It looks like zero and negative zero are not handled by the outer code (just NaN/Inf).

mazong1123 replied:
Yes, we should shortcut zeros here. Thanks. I'd like to borrow the code from https://github.com/dotnet/coreclr/blob/master/src/pal/src/cruntime/misctls.cpp#L185-L193

However, I do not quite understand the comment at https://github.com/dotnet/coreclr/blob/master/src/pal/src/cruntime/misctls.cpp#L184
> // we have issue #10290 tracking fixing the sign of 0.0 across the platforms

Did this code introduce #10290 issue?

tannergooding replied:
@mazong1123, no. #10290 is tracking a future improvement (and possible breaking change) that -0.0 and 0.0 should be treated differently when converting to a string.

The code you referenced is fine and is matching the current expected behavior.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    UINT64 realMantissa = 0;
+    int realExponent = 0;
+    UINT32 mantissaHighBitIdx = 0;
+    if (((FPDOUBLE*)&value)->exp > 0)
```
I would expect you to be special casing denormal numbers, since the implicit bit is zero instead of one.

mazong1123 replied:
I think I have included the denormal numbers in the else condition. According to https://en.wikipedia.org/wiki/Double-precision_floating-point_format 
> In the case of subnormals (e=0) the double-precision number is described by:
(-1)^sign * 2^(1 - 1023) * 0.fraction = (-1)^sign * 2^(-1022) * 0.fraction

2^(-1022) * 0.fraction = 2^(-1022) * mantissa / 2 ^52 * 2^(-1022) = mantissa * 2^(-1074) . So you can see in the `else` block, `realExponent = -1074` , `realMantissa = mantissa`.

tannergooding commented on src/classlibnative/bcltype/bignum.h:
```diff
@@ -0,0 +1,140 @@
+        }
+    } m_initializer;
+
+    UINT8 m_len;
```
nit: Either make this UINT32 or place it after `m_blocks` (larger fields should come first). The resulting struct is going to be 8-bytes in either case...

mazong1123 replied:
I'll place `m_len` after `m_blocks`. Thanks.

mikedn replied:
You really should make it UINT32 rather than placing it after `m_blocks`. In general, the only advantage of using small integral types is storage size. In this case you can't actually save any storage due to alignment requirements. Instead your risk making the code larger because the compiler may need to insert widening conversion in some cases.

mazong1123 replied:
@mikedn  Sounds reasonable. Thanks.

mikedn replied:
> Something like BIGSIZE and BIGPOWER10NUM should be extended to UINT32.

Yes, but only for consistency. I would expect the compiler to treat those as constants so the impact on the generated code should be exactly 0. In fact, they should be `constexpr`, provided that the version(s) of VC++ used to build this code doesn't barf on it.

tannergooding commented on src/classlibnative/bcltype/bignum.h:
```diff
@@ -0,0 +1,140 @@
+    static const UINT8 BIGSIZE = 35;
+    static const UINT8 UINT32POWER10NUM = 8;
+    static const UINT8 BIGPOWER10NUM = 6;
+    static UINT32 m_power10UInt32Table[UINT32POWER10NUM];
```
static UINT32 const m_power10UInt32Table[] =
{
    // Declare data inline
};

mazong1123 replied:
I have to use `constexpr` to accomplish this. The code can be compiled on my current Ubuntu 16.04 box but I'm not sure if it is able to be compiled across platform. I'll do the change and to see if it can pass the CI later.

tannergooding replied:
Using `constexpr` would likely work, but it shouldn't be required.

This should just be a `UINT32` constant for the length and then a sequence of bytes representing the data. You likely need a second lookup table that contains the starting index for each entry, but it should just be a cast from UINT32* to BigNum* at that point.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+
+BigNum::BigNum()
+    :m_len(0)
+{
```
a comment about explicitly not zeroing the memory due to perf reasons might be nice

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+            break;
+        }
+
+        --pCurrentLeftBlock;
```
nit: Lengths of both are the same, you can use a single index to do the access

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+    return *this;
+}
+
+int BigNum::Compare(const BigNum& lhs, UINT32 value)
```
Is having a `compare` really better than individual ==, !=, <, <=, >, and >= operators?

mazong1123 replied:
I can overload these operators but I think we should do this enhancement in another PR because a simple compare method should be enough to implement the algorithm.

tannergooding replied:
I was mostly asking from a perf perspective. It seems like we are only actually reusing the Compare value in one place, every other place we are just doing a single comparison, which is likely cheaper than the full Compare check.

mazong1123 replied:
I'll remove `BigNum::Compare(const BigNum& lhs, UINT32 value)` since it is not used now. 

`BigNum::Compare(const BigNum& lhs, const BigNum& rhs)` is still useful when comparing numerator and denominator so I'll keep it.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+    int shiftBlocks = shift / 32;
+    int remaningToShiftBits = shift % 32;
+
+    for (int i = 0; i < shiftBlocks; ++i)
```
memset(0) would be much faster

tannergooding replied:
(memset over the entire range that is)

mikedn replied:
Might be, it really depends on how large these bignums usually get. I suspect not very large.

tannergooding replied:
As per the `memcpy` comment. `memset` should do the right thing, regardless of size. The current max size is only 35 bytes, but I also believe this is considerably too small.

mazong1123 replied:
I'll change the loop to memset.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+        return true;
+    }
+
+    for (UINT8 i = 0; i < m_len; ++i)
```
it seems like we could trivially detect all the cases where the value would become zero and only rely on m_len == 0

mazong1123 replied:
Did you mean we should ensure that whenever `m_len > 0`, we at least have 1 block != 0 so that we could only check `m_len == 0` inside IsZero()?

tannergooding replied:
IsZero is part of the `BigNum` class, so it would only be dependent on a convention set for its own class.

This probably isn't too important so we can investigate later.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+
+void BigNum::SetUInt64(UINT64 value)
+{
+    m_len = 0;
```
m_blocks[0] = value & 0xFFFFFFFF;
m_blocks[1] = value >> 32;
m_len = (m_blocks[1] == 0) ? 1 : 2;

mazong1123 replied:
Nice! Thanks!

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+
+void BigNum::ExtendBlock(UINT32 newBlock)
+{
+    m_blocks[m_len] = newBlock;
```
There are a few cases where an `ExtendBlock(UINT32 blockValue, UINT32 numBlocks)` would be useful. It can `memset` and increment m_len the appropriate count all at once.

mazong1123 replied:
I tried this change but it slightly slows down the performance. So I'd like to keep this method and use memset when there're bulk blocks need to be extended.

tannergooding replied:
Did you create a second `ExtendBlock` method or just replace this one entirely? I was suggesting that we create a second method and call it if the number of blocks being set (we loop the current method in a couple places) is larger than some threshold.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    // To keep the precision, we represent the double value as numertor/denominator.
+    BigNum numerator;
+    BigNum denominator;
+    if (realExponent > 0)
```
isn't this supposed to be >= 0, based on the paper?

mazong1123 replied:
The `realExponent` represents as the `e` in `f * b^e` in the paper, which is `be - 1075` for normalized value or `-1074` for denormalized value.  I think either `>` or `>=` here will produce correct `numerator` and `denominator`, if you go through the code block.

tannergooding replied:
I was specifically referring to the initialization of `numerator` (r) and `denominator` (s) according to this:

>  Initialize r, s, m+, and m− such that v = r s, v−v− 2 = m− s , and v+−v 2 = m+ s according to Table 1.

#### Table 1
![image](https://user-images.githubusercontent.com/10487869/28397240-4165010c-6cb5-11e7-88e4-c7f3379292d9.png)

mazong1123 replied:
Got it. Although the result is the same, I'd better follow the table here. So I'll change `> 0` to `>= 0`.

tannergooding replied:
I wouldn't expect the result to be the same... The exact handling for `e >= 0` is meant to vary based on whether or not `f = b^(p-1)`.

mazong1123 replied:
Well, the result should be the same IMO.

Let's say what would happen if we process 0 with the `e < 0` logic:
### When f <> b^(p - 1):
r = f * 2
s = b^-e * 2 = 2

### When f = b^(p - 1), 
r = f * b * 2 
s = b^(-e + 1) * 2 = b * 2

The result if we process 0 with e > 0 logic:
### When f <> b^(p - 1):
r = f * b^e * 2 = f * b^0 * 2 = f * 2;
s = 2;

### When f = b^(p - 1), 
r = f * b^(e + 1) * 2 = f * b ^1 * 2 = f * b * 2;
s = b * 2;

So we got the same result :)

tannergooding replied:
You're right :). Please document why we deviated (or match the table and ensure we optimize `multiply by 1` in the BigInt implementation).

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    // Store the input double value in BigNum format.
+    //
+    // To keep the precision, we represent the double value as numertor/denominator.
+    BigNum numerator;
```
nit: numerator is always `realMantissa`, set it once before the condition

mazong1123 replied:
Thanks. I'll change it.

tannergooding commented:
General feedback: The long names (such as `realMantissa`) make it very hard to validate that the implementation follows the algorithm described in the paper (where they use `r`, for example).

It would be much easier to follow/validate if you used the variable names declared in the paper and had a brief excerpt at the top defining each variable.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+        BigNum::ShiftLeft(&numerator, realExponent);
+
+        // Explanation:
+        // value = (realMantissa * 2^realExponent) / (1)
```
This doesn't seem to be handling `f != b^(p−1)` vs `f = b^(p−1)`

mazong1123 replied:
I think regardless `f != b^(p - 1)` or `f = b^(p - 1)`, `r / s` can be reperesented as `f * b^e`

When `f != b^(p - 1)`:
r: f * b^e * 2
s: 2
r / s = f * b^e

When `f = b^(p - 1)`
r: f * b^(e + 1) * 2
s: b * 2
r / s = f * b^e

I'll document this in the code.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+        BigNum::ShiftLeft(1, -realExponent, denominator);
+    }
+
+    if (firstDigitExponent > 0)
```
I would have expected this to be >= 0, based on the paper

mazong1123 replied:
Actually this is a performance improvement. We do not need to do any operation on `numerator` and `denominator` when firstDigitExponent is 0. The corresponding code in the paper is:

```
(if (>= est 0)
(fixup r (∗ s (exptt B est)) m+ m− est B low-ok? high-ok? )
(let ([scale (exptt B (− est))])
(fixup (∗ r scale) s (∗ m+ scale) (∗ m− scale) est B low-ok? high-ok? ))))

```

If`est` is 0, `(∗ s (exptt B est)) = s`,  `(∗ r scale) = (* r (exptt B (− est)))) = r`.

tannergooding replied:
Please document the deviation and why.

If I simplified correctly, this is correct and the algorithm does essentially become `denominator *= 10^firstDigitExponent` which is `denominator *= 1` for `firstDigitExponent == 0`

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    if (firstDigitExponent > 0)
+    {
+        BigNum poweredValue;
+        BigNum::Pow10(firstDigitExponent, poweredValue);
```
I'm pretty sure multiplying by pow10 can be made significantly cheaper than doing an actual multiply...

mazong1123 replied:
I'm also thinking about this. Currently the pow10 still requires a big integer multiply if the exp > 7. That's not good.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    // Output digits until reaching the last but one precision or the numerator becomes zero.
+    int digitsNum = 0;
+    int currentDigit = 0;
+    while (true)
```
A UINT32 can represent a value of up to 10^9, so you only need `big_int_val % 9` big integer divisions (and therefore iterations), the rest can be regular UINT32 divisions.

tannergooding commented:
still reviewing, but need to take a break for a bit. Will pick back up after dinner.

mazong1123 commented:
@tarekgh it's worth mentioning that some of the ideas and code were modified from http://www.ryanjuckett.com/programming/printing-floating-point-numbers/ , especially for how to estimate the first digital exponent and doing a heuristic divide. If they are not good to be adopted to .NET core, I need to remove them.

mazong1123 commented:
@tannergooding thanks for your feedback. I'll go through them tonight (in China). Thanks!

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
 
-#pragma warning(disable:4273)
-extern "C" char* __cdecl _ecvt(double, int, int*, int*);
+void ecvt( double value, int count, int* dec, int* sign, wchar_t* digits )
```
Is this replacing the Windows implementation as well?

mazong1123 replied:
I thought the Windows implementation is in elsewhere: https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/vm/i386/fptext.asm#L78

Are you saying delete Line 137 - 138 will force Windows implementation being replaced by mine?

tannergooding replied:
That is only the 32-bit implementation which does not hit this particular code path.

This particular _ecvt declaration should be hit if:
* `_TARGET_X86_` is not defined
or
* `_TARGET_X86_` and `FEATURE_PAL` are both defined

So 64-bit windows and all non-windows platforms.

There still might be some code I'm missing somewhere, but we should make sure that is the case. We likely don't want to replace the windows implementation where the code is likely still faster than the replacement here.

mazong1123 replied:
If I understand correctly, the `FEATURE_PAL` can indicate whether the OS is Windows or not. If that is correct, I think ifdef out the code like this should resolve the issue:

```cpp
#ifndef  FEATURE_PAL
#pragma warning(disable:4273)
extern "C" char* __cdecl _ecvt(double, int, int*, int*);
#else
void ecvt( double value, int count, int* dec, int* sign, wchar_t* digits )
...
...
#endif
```

And on [line 336](https://github.com/dotnet/coreclr/pull/12894/files/b617b7928f47263980752318951663b7faa79dee#diff-0d0ac521b3af6ae20f3148fa1bee21f1L152):

```cpp
#ifndef FEATURE_PAL
 char* src = _ecvt(value, precision, &number->scale, &number->sign);
 wchar* dst = number->digits;
 if (*src != '0') {
      while (*src) *dst++ = *src++;
 }
 *dst = 0;
#else
ecvt(value, precision, &number->scale, &number->sign, number->digits);
#endif
```

tannergooding replied:
That would probably work. Although, just calling it _ecvt everywhere should too (this might require a minor cleanup in the PAL layer where the other non-windows implementation of _ecvt is)

tannergooding replied:
Also, regardless of which route you decide. It would probably be good to remove the old _ecvt implementation (defined in the PAL layer) anyways, as it is dead code now.

tarekgh replied:
I think the current change is enabling the new implementation for none x86 compilation. I mean your changes should be reflected on amd64 Windows compilation too.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    }
+    else
+    {
+        numerator.Multiply(10);
```
Multiplication by 10 can be simplified to: `(x << 3) + (x << 1)`
`(x << 3) + (x << 1)` becomes `(x * 8) + (x * 2)` becomes `8x + 2x` becomes `10x`

This is likely faster than doing actual big integer multiplication

mazong1123 replied:
I think it may not gain much. If I understand correctly, we need 2 big integer shift operations and 1 big integer add operation, probably need a temporary memory to store one of the shift results. That sounds slower than just do a big integer multiplication.

tannergooding replied:
It's either 2 shifts and an add or 1 shift and two adds (`x * 2` can be either `x << 1` or `x + x`).

We can probably investigate later, but I would think that this would be faster (a lot of these math optimizations need just a single temporary `BigNum`, it would add additional memory overhead but would probably provide measurable savings -- BigNum shift and add is often way cheaper than BigNum multiply).

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    //  = compare( numerator / denominator, 0.5 )
+    //  = compare( numerator, 0.5 * denominator)
+    //  = compare(2 * numerator, denominator)
+    numerator.Multiply(2);
```
`numerator = numerator + numerator` or `numerator = numerator << 1` will likely be faster and achieve the same thing.

mazong1123 replied:
Nice. I'll change this to `numerator.ShiftLeft(1)`.

mazong1123 commented:
> General feedback: The long names (such as realMantissa) make it very hard to validate that the implementation follows the algorithm described in the paper (where they use r, for example).

It would be much easier to follow/validate if you used the variable names declared in the paper and had a brief excerpt at the top defining each variable.

How about do it reversely? I keep current meaningful variable name, and add a brief excerpt to show the relationship between the meaningful variable name and the name in the paper? Then we would not force the developer to read the paper before read the code (if they do not want to go deeper).

tannergooding commented:
> How about do it reversely? I keep current meaningful variable name, and add a brief excerpt to show the relationship between the meaningful variable name and the name in the paper? Then we would not force the developer to read the paper before read the code (if they do not want to go deeper).

Someone else should chime in (@tarekgh), but I am of the opinion that: When implementing an algorithm based on a paper, the following should be done:
* Note the paper being used with a brief summary of its benefts/drawbacks at the top of the method
* Implement the algorithm as closely to the paper as possible, explicitly noting *any* deviations (even simple optimizations)
  * The paper has likely undergone various scrutiny (especially algorithms that have been in use for a long time, like this one), any optimizations being made have not
* Use variable names that match the paper whenever possible
  * This also comes out from the second point above. Anyone updating this implementation needs to understand that it is following an existing paper and that they should be documenting any deviations/etc. This also means they should have a rough understanding of why things are done the way they are before making modifications (they should read the paper, which isn't very long, before changing the code).

tannergooding commented on src/classlibnative/bcltype/bignum.h:
```diff
@@ -0,0 +1,140 @@
+
+private:
+
+    static const UINT8 BIGSIZE = 35;
```
This is significantly smaller than I would expect is required for some inputs... Just representing 2^1074 + shift space be 138 bytes.

tannergooding replied:
In order to represent `s` in the below table, we need to be able to represent the largest value for b^-e at a minimum. b is always 2 and the maximum real e is 1074. This means we need at least 1074 bits or 135 (134.25) bytes. We additionally need space for the maximum shift (32-bits or 4 bytes).

Thus, at a minimum, we should need at least 139 bytes. On top of this, we likely need space for the maximum number of decimal digits we will process which comes out to `log2(10^max_digits)` additional bits (where max_digits depends on the buffer size).

![image](https://user-images.githubusercontent.com/10487869/28397240-4165010c-6cb5-11e7-88e4-c7f3379292d9.png)

mazong1123 replied:
I may misunderstand something but `UINT32 m_blocks[BIGSIZE]` should be able to represent 2^(35 * 32) = 2^1120 . Did you mean something else?

tannergooding replied:
Sorry, this was my mistake... I was reading this in `bytes` when I should have been reading it in `dwords` (currently it is 140 bytes).

However, just initializing the smallest denormal value will fill almost every byte, are we sure that 5 bytes is sufficient for all other operations?

tarekgh replied:
> Someone else should chime in (@tarekgh), but I am of the opinion that: When implementing an algorithm based on a paper,

I prefer to use the same variable used in the paper and then have a comment section in the code describing the meaning of each variable. The only challenge here is m+ and m- so I suggest to use mp and mm for it.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+
+#include "bignum.h"
+
+UINT32 BigNum::m_power10UInt32Table[UINT32POWER10NUM] =
```
You can do up to 10^9 in a single `UINT32`

tannergooding commented on src/classlibnative/bcltype/bignum.h:
```diff
@@ -0,0 +1,140 @@
+    public:
+        StaticInitializer()
+        {
+            // 10^8
```
Why by powers of 2? (8, 16, 32, 64, etc)

mazong1123 replied:
This is a little bit tricky. It's along with m_power10UInt32Table to speed up BigNum::Pow10(exp, result). If we have a more efficient implementation I can remove this one.

Let's say we pass a value whose binary representation is `111111` to `BigNum::Pow10(exp, result)`.

We can split the exp to two parts, one is small exp, which 10^smallExp can be represented as UINT32, another part is 10^bigExp, which must be represented as BigNum. So the result should be `10^smallExp * 10^bigExp`.

Calculate 10^smallExp is simple, we just lookup the 10^smallExp from `m_power10UInt32Table`. But here's a bad news: Although UINT32 can represent 10^9, exp 9's binary representation is `1001`. That means 10^(1011), 10^(1101), 10^(1111) all cannot be stored as UINT32, we cannot easily say something like: "If exp <= 3 bits, its a smallExp, if exp > 3 bits, its a bigExp". So instead of involving 10^8, 10^9 to m_power10UInt32Table, I consider 10^8, 10^9 as a bigNum, so they fall into `m_power10BigNumTable`. Now we can have a simple rule: "If exp <= 3 bits, its a smallExp, if exp > 3 bits, its a bigExp".

For (111111)2, we first calculate 10^(smallExp), which is 10^(7), now we can shift right 3 bits, prepare to calculate the bigExp part, the exp now becomes (000111)2. 

Apparently the first bit of bigExp should reperesent 10^8 because we have already shifted 3 bits for smallExp, so m_power10BigNumTable[0] = 10^8.
Now let's shift exp right 1 more bit, the second bigExp bit should represent 10^(8 * 2) = 10^16. Then so on...

That's why we just need the value of m_power10BigNumTable be power of 2.

This idea is from http://www.ryanjuckett.com/programming/printing-floating-point-numbers/ . I think it's tricky but faster than a raw pow10 implementation. I'm still looking for a more efficient and elegant pow10 implementation. If there is, I'd like to replace this one.

tannergooding replied:
So I understand why we have the gaps (it saves space and lookup is somewhat trivial) and that we essentially interpolate between the two tables to get missing values.

I don't understand the reasoning for `if exp <= 3 bits`. If you did up to 9 in the small table, then you could go by powers of 10 (and still have the same lookup/interpolation).

In either case, you essentially have `smallExp = exp % entriesInSmallTable`, `bigExp = exp - smallExp`. Then `pow10 = smallTable[smallExp] * smallTable[bigEx]`

Actually, I may have just explained it to myself. Division/Remainder for powers of 2 is cheaper than Division/Remainder for powers of 10, so the code is likely slightly more performant.

In any case, documenting the reasoning (for future reference) is likely beneficial.

tannergooding commented:
You need to test the minimum subnormal positive double (`BitConverter.Int64BitsToDouble(1)`).

tannergooding commented:
You should also test that `-0` and `+0` are being handled correctly. As well as `Inf`, `-Inf`, and `NaN` (if these are handled before the method, then assert that the value is never one of these).

mazong1123 commented:
> You should also test that -0 and +0 are being handled correctly. As well as Inf, -Inf, and NaN (if these are handled before the method, then assert that the value is never one of these).

These should already be handled before calling the method: https://github.com/dotnet/coreclr/pull/12894/files/b617b7928f47263980752318951663b7faa79dee#diff-0d0ac521b3af6ae20f3148fa1bee21f1L146

I'll double confirm them. Add asserts is a good choice - I'll do that.

tannergooding commented on src/classlibnative/bcltype/bignum.h:
```diff
@@ -0,0 +1,140 @@
+    } m_initializer;
+
+    UINT8 m_len;
+    UINT32 m_blocks[BIGSIZE];
```
It would be good to have asserts around `m_len` and around indexes into `m_blocks` to ensure that we aren't accidentally overflowing anywhere.

mazong1123 replied:
Sure. I'll do that.

tarekgh commented:
We may need to delete the implementation in https://github.com/dotnet/coreclr/blob/master/src/pal/src/cruntime/misctls.cpp#L156

tarekgh commented:
Why did we choose implementing https://www.cs.indiana.edu/~dyb/pubs/FP-Printing-PLDI96.pdf and not Dragon 4 algorithm http://www.ryanjuckett.com/programming/printing-floating-point-numbers/part-2/? I expect Dragon 4 is more recent and should be performant too (I guess).

tannergooding commented:
@tarekgh, I think its probably worth just doing `PLDI96` to start (with no major deviations). It is a known good algorithm that is already used in other major CRT implementations.

A secondary PR that investigates further improvements (based on other/newer papers) could be done after the fact.

mazong1123 commented:
> Why did we choose implementing https://www.cs.indiana.edu/~dyb/pubs/FP-Printing-PLDI96.pdf and not Dragon 4 algorithm http://www.ryanjuckett.com/programming/printing-floating-point-numbers/part-2/? I expect Dragon 4 is more recent and should be performant too (I guess).

Dragon 4 is the name of the algorithm on the paper. The author of the website just implemented the algorithm on the paper with two major improvement, that is:

1. Reduce the chance to estimate the wrong first digital exponent.
2. A heuristic divide to improve the performance of big integer division.

I adopted these two improvements and commented in my code.

tarekgh commented:
@mazong1123 so we basically picked the implementation from http://www.ryanjuckett.com/programming/printing-floating-point-numbers/part-2/ right?

I am asking because I am seeing the following comment on the code and I am currently communicating to the LCA regarding the license of using the algorithm and code. we need to be clear wht exactly we used. 

```C
/******************************************************************************
  Copyright (c) 2014 Ryan Juckett
  http://www.ryanjuckett.com/
 
  This software is provided 'as-is', without any express or implied
  warranty. In no event will the authors be held liable for any damages
  arising from the use of this software.
 
  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:
 
  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
 
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
 
  3. This notice may not be removed or altered from any source
     distribution.
******************************************************************************/
```

mazong1123 commented:
@tarekgh The directly used code:

How to estimate the first digit exponent:
```cpp
int firstDigitExponent = (int)(ceil(double((int)mantissaHighBitIdx + realExponent) * 0.30102999566398119521373889472449 - 0.69));
```
The implementation of heuristic divide:
```cpp
BigNum::HeuristicDivide()
```
I think these two are the key work of the author of the website and I used them directly. 

His code supports both free format input and fixed format input. I leveraged his fixed format input idea along with the paper to implement the ecvt which only supports fixed format input.

I think you have to contact the author because I ignored the License you listed when I wrote the code. If it's a problem I need to remove these two major improvements created by the author, and get rid of any idea from his website (It's a nice implementation though).

tarekgh commented:
@mazong1123 thanks for the clarification. what about the rest of the big number implementation? I am following about with LCA anyway and communicated that we are using the paper and also ideas from  http://www.ryanjuckett.com/programming/printing-floating-point-numbers/part-2/

Could you please fix the code according to the review comment so we can measure the perf on Windows too? thanks again for your effort here.

mazong1123 commented:
The BigNum implementation is hard to say. I do get some ideas from his code, but except the heuristic dividing, big integer operations are widely used common technique, it's likely neither of us are the original author of them:)

I'm fixing the code regarding the feedback. Regarding testing on Windows, I think x86 implementation will not fall into the code path I implemented. We can only test x64 Windows here.

tarekgh commented:
>  I think x86 implementation will not fall into the code path I implemented. We can only test x64 Windows here.

That is right. we can start measuring 64 bit first and then we can look at x86 later.

janvorli commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,597 @@
+    ++m_len;
+}
+
+UINT32 BigNum::LogBase2(UINT32 val)
```
How about using `__builtin_clz` for FEATURE_PAL and `_BitScanReverse` / `_BitScanReverse64` for Windows? My guess is that it would be the fastest way to compute the logarithm.

mazong1123 replied:
I'd like to give `__builtin_clz` a shot and measure the performance later.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -137,86 +137,185 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
 {
+    // ========================================================================================================================================
+    // This implementation is based on the paper: https://www.cs.indiana.edu/~dyb/pubs/FP-Printing-PLDI96.pdf
+    // Besides the paper, some of the code and ideas are modified from http://www.ryanjuckett.com/programming/printing-floating-point-numbers/
```
not positive, but I'm fairly certain we need to preserve the license header listed by the site.
@tarekgh can confirm (and follow up with LCA if necessary).

mazong1123 replied:
I'm waiting for @tarekgh's feedback. Either add the header or we simply cannot use the code.

mazong1123 commented:
Seems like constexpr introduced a compile issue on Windows. I'm going to figure it out.

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -137,86 +137,185 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
-        realMantissa = ((UINT64)(((FPDOUBLE*)&value)->mantHi) << 32) | ((FPDOUBLE*)&value)->mantLo;
-        realExponent = -1074;
-        mantissaHighBitIdx = BigNum::LogBase2(realMantissa);
+        // For denormalized value, according to https://en.wikipedia.org/wiki/Double-precision_floating-point_format
```
I don't think this is just for `denormalized` values which is only `exp = 0`. Based on the above comparison for `exp > 0`, this will also cover normalized values with a negative exp.

mazong1123 replied:
For normalized values, the exp here should be in 1 ~ 2046. A `biased exp` which in the paper is `e` can be negative.

tannergooding replied:
If that is the case, a comparison of `!= 0` would cause less confusion.

mazong1123 replied:
Make sense. I'll.change it. Thanks

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -137,86 +137,185 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    // Following are the Scheme code from the paper:
+    // --------------------------------------------------------------------------------
+    // (if (>= est 0)
+    // (fixup r (∗ s (exptt B est)) m+ m− est B low-ok? high-ok? )
```
It would be great to document why we don't use `m+` and `m-` in our own code 😄

mazong1123 replied:
I'll.document it. Thanks.

mazong1123 commented:
@tannergooding I added a new Multiply10() method to specialize the multiplication by using shift and add operations. It does speed up 10% performance.

@tannergooding @tarekgh I've removed `_ecvt` implementation and tests from `pal`. However, I think the test cases are good to apply for the new ecvt implementation. Do you have any idea where can I add the tests for the new ecvt? I don't find any existing tests for the `bcltype` module.

janvorli commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,608 @@
+    _ASSERTE(value != 0);
+
+#if defined(FEATURE_PAL)
+    return (UINT32) (8 * sizeof (UINT32) - __builtin_clz(value) - 1);
```
Rather than using the `__builtin_clz` and `__buildtin_clzll` functions  here and having to have the ifdef, please add implementation of `_BitScanReverse` and `_BitScanReverse64` to pal.h instead. There are already implementations for the forward variants of the bit scan there, so please put these next to those.

janvorli commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,608 @@
+#include "bignum.h"
+
+#if !defined(FEATURE_PAL)
+#include <intrin.h>
```
Please make this include unconditional. We have an `intrin.h` for `FEATURE_PA`L which ends up including `pal.h`.

mazong1123 commented:
@janvorli I've updated according to your feedback. Please take a look. Thanks!

janvorli commented:
@mazong1123 the updated version looks good, thank you!

mazong1123 commented:
@tarekgh  @tannergooding  Now the new implementation is about 4x faster than the old one on Ubuntu 16.04. Since I don't have a Windows box on hand, could you please do me a favor to test it on Windows? Thanks!

tarekgh commented:
@mazong1123 thanks for getting this ready. I'll try to get into measuring the perf on Windows by tomorrow or so.

mazong1123 commented:
@dotnet-bot test Windows_NT x64 Debug Build and Test  please

tarekgh commented:
I have measured the perf on Windows and I am seeing reasonable improvement too. I have used the exact same test that @mazong1123 used in the measurement on Linux. and here is some numbers:

Running the test on a simple double number ```104234.343 ```

Before the change  | After the change
------------ | -------------
  42 | 36
  40 | 38
  39 | 37
  39 | 35
  39 | 35
  39 | 37
  40 | 37
  39 | 36
  39 | 35
  39 | 35
  39 | 35
  40 | 36
  40 | 36
  39 | 37
  39 | 36
  39 | 37
  39 | 36
  39 | 36
  48 | 36
  74 | 36

In average there is improvement by ```13%```

Running the test on a complex double number ```double.MinValue / 2``

Before the change  | After the change
------------ | -------------
  422 | 136
  385 | 132
  386 | 142
  386 | 135
  384 | 134
  384 | 134
  384 | 134
  396 | 134
  386 | 143
  392 | 133
  385 | 134
  385 | 133
  386 | 133
  386 | 133
  385 | 134
  385 | 134
  384 | 133
  385 | 136
  394 | 145
  391 | 192

In average there is improvement by ```64%```

The changes look good enough for both Windows and Linux. 

@mazong1123 is there any pending changes from your side? I am still following up with the LCA so this may take some time but I think we are getting close. and thanks for getting this implemented with this shape. 

@jkotas do you want to code review this one?

mazong1123 commented:
Do we have a place to add tests for the new implementation? I think the tests for the old _ecvt are good for the new one. But I don't find a place to put them.

tarekgh commented:
> Do we have a place to add tests for the new implementation? I think the tests for the old _ecvt are good for the new one. But I don't find a place to put them.

Usually, we add the tests in corefx repo. so you can add the test there but you have to wait to enable this test till we get your changes merged to coreclr and then corefx consume the new coreclr package.

If you want to work writing the test on corefx from now, you can add the test to the file https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/DoubleTests.cs

To consume your private built coreclr with your changes in corefx, you will need to build coreclr and then when building corefx use the command 

```
./build.sh -- /p:CoreCLROverridePath=<path to your private coreclr>
```

tannergooding commented:
It would also be worth validating perf for a range of other numbers (denormals, 0, -0, NaN, +/- Infinity, Pi, etc).

tarekgh commented:
Here are some measurement results for different numbers on Windows. Time measured by ms

Number          | Avarage time before | Avarage time after | delta percentage
------------ | ------------- | ------------- | -------------
104234.343      | 41  | 34  | 17%
double.MinValue | 417 | 129 | 69%
PI              | 43  | 29  | 33%
E               | 43  | 29  | 33%
MaxValue        | 396 | 136 | 66%
MinValue        | 394 | 136 | 66%
NaN             | 1   | 1   | 0%
-∞              | 1   | 1   | 0%
∞               | 1   | 1   | 0%
Max Subnormal   | 439 | 128 | 71%
Min Subnormal   | 430 | 132 | 69%
middle Subnormal| 434 | 132 | 70%
0.0             | 11  | 7   | 36%
-0.0            | 12  | 7   | 42%

tarekgh commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
On Linux, each test case on new implementation are much faster than or equal to the old one except `0.0`. New implementation is -40% slower than the old one. Below are the perf table:

Number | Avarage time before | Avarage time after | delta percentage
-- | -- | -- | --
104234.343 | 170 | 49 | 71%
double.MinValue/2 | 382 | 178 | 53%
PI | 67 | 45 | 33%
E | 67 | 44 | 34%
MaxValue | 365 | 193 | 47%
MinValue | 365 | 194 | 47%
NaN | 3 | 3 | 0%
-∞ | 3 | 3 | 0%
∞ | 3 | 3 | 0%
Max Subnormal | 126 | 52 | 58%
Min Subnormal | 135 | 62 | 54%
middle Subnormal | 126 | 52 | 58%
0.0 | 10 | 14 | -40%
-0.0 | 135 | 62 | 54%

I'm trying to introduce `wmemset` to improve the performance. However, it seems like `Tizen armel` tests failed by this change. I guess I need to work another way to improve the performance of converting `0.0` on Linux.

jkotas commented on src/pal/inc/pal.h:
```diff
@@ -5240,6 +5286,7 @@ PALIMPORT void * __cdecl memcpy(void *, const void *, size_t);
 #endif //defined(_DEBUG)
 PALIMPORT int    __cdecl memcmp(const void *, const void *, size_t);
 PALIMPORT void * __cdecl memset(void *, int, size_t);
+PALIMPORT wchar_t * __cdecl wmemset(wchar_t *, wchar_t, size_t);
```
I do not think that this works. CoreCLR redefines `wchar_t` as 16-bit type, but the system has `wchar_t` defined as 32-bit type.

mazong1123 replied:
Yep. I gave up the `wmemset` thing. Trying to find another way to improve the perf of converting `0.0`. I found I'm using almost the same code as old implementation except the old one assign '0' to the LPSTR* array, whereas I'm assigning L'0' to wchar_t* array.
https://github.com/dotnet/coreclr/blob/master/src/pal/src/cruntime/misctls.cpp#L187-L190

tannergooding replied:
Any reason why we can't use `wmemset` on windows and the copy loop on everything else (with a bug tracking an eventual fix)?

mazong1123 commented:
@dotnet-bot test Ubuntu arm Cross Release Build please.

tannergooding commented:
@mazong1123, @tarekgh. For the basic perf tests you ran above, what was the requested precision? Did you just do `number.ToString()` or did you also try it with varying precision.

The default precision being performant is great, but we should also ensure that other precisions are remaining performant as well (7, 9, 15, and 17 are the ones we have specific code-paths for).

That being said, we probably also want to test other values like `ECVT_MAX_BUFFER_SIZE` (or whatever the max buffer size of `number->digits` is).

mazong1123 commented:
@tannergooding Seems like we may need to write an official perf test program other than the microbench. Do you guys have any tool or framework for writing perf tests? Or I need to write from scratch?

mazong1123 commented:
@dotnet-bot test Ubuntu arm Cross Release Build please

tannergooding commented:
You can look at https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality/Math/Functions for an example of a perf test on the CoreCLR side.

I'm not sure about the CoreFX side of things.

Someone else may want to comment on whether or not this is the recommended way to add a perf test, or if there is a better process now 😄

mazong1123 commented:
@dotnet-bot test this please

tarekgh commented:
@mellinoe could you please advise what is the best writing a perf test which we can use on Windows and Linux?

mellinoe commented:
We have some documentation about adding performance tests to CoreFX in [this document](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/performance-tests.md#writing-test-cases) Some of the sections before the one I linked are a little bit out of date, but they are describing how to set up a new test project, which you hopefully will not need to do. The section I linked describes how to write the actual test code, which is the important part.

mazong1123 commented:
@mellinoe Thanks! I'm gonna write the perf tests based on the document. 

@tarekgh if I understand correctly, I cannot directly add the perf tests to CoreFX repo because the new CoreCLR changes have not been adopted. So I'll create a new repository to store the perf tests. After CoreFX adopt this change, I'll add the perf tests to the codebase.

mazong1123 commented:
@tannergooding @janvorli I've changed the function name from `ecvt` to `DoubleToNumberWorker`. Could you please take a look? Thanks!

mazong1123 commented:
I've written perf tests at https://github.com/mazong1123/doubleperftest

I tested the precision from G0 to G17 and R with the numbers listed previously.

Generally speaking, on Ubuntu 16.04, the new implementation is really much faster than the old one. The detail results can be found in the attachment (it's too big to show within comments...)

@tarekgh I think you may use my perf test code to measure the performance on Windows :)

[newresults.zip](https://github.com/dotnet/coreclr/files/1179613/newresults.zip)
[oldresults.zip](https://github.com/dotnet/coreclr/files/1179612/oldresults.zip)

tarekgh commented:
> @tarekgh if I understand correctly, I cannot directly add the perf tests to CoreFX repo because the new CoreCLR changes have not been adopted. So I'll create a new repository to store the perf tests. After CoreFX adopt this change, I'll add the perf tests to the codebase.

You can add the perf test from now without waiting for your change to come. when you add these perf tests will be exercised with the old code and we'll get the numbers before your changes. later when we have your changes, the tests will start producing the new numbers and we can compare it with the old results and know the exact perf gain numbers. so please go ahead and add such tests. Also, my understanding is writing the test will work for both Windows and Linux. so we don't have to write specific tests for Windows and others for Linux.

@mellinoe please correct me if I say anything wrong here.

tarekgh commented:
@mazong1123 are you willing to add the perf tests? just checking to plan what is next need to be done. 

For the current PR, we are still waiting for the LCA.

mazong1123 commented:
@tarekgh The perf tests have already been merged into CoreFX https://github.com/dotnet/corefx/pull/22714

:)

tarekgh commented:
I missed that. Even didn't reference this PR ;-)

Thanks @mazong1123, this great!

tarekgh commented:
good news, we got the green light from the LCA and we are good here. 

@jkotas @AlexGhiondea @joperezr @tannergooding are you good with the changes? can we merge it?

tannergooding commented:
I'll give a final review pass when I have a free moment (probably tonight).

mazong1123 commented:
Do I need to put the license mentioned in https://github.com/dotnet/coreclr/pull/12894#issuecomment-316875853 anywhere?
At the beginning of number.cpp or DoubleToNumberWorker?

tarekgh commented:
> Do I need to put the license mentioned in #12894 (comment) anywhere?
At the beginning of number.cpp or DoubleToNumberWorker?

I am asking and will get back to you.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,599 @@
+#include <intrin.h>
+
+constexpr UINT32 BigNum::m_power10UInt32Table[UINT32POWER10NUM];
+BigNum BigNum::m_power10BigNumTable[BIGPOWER10NUM];
```
is this meant to be `constexpr` as well?

mazong1123 replied:
`m_power10BigNumTable` needs to be initialized in https://github.com/mazong1123/coreclr/blob/9b39ea6c0e28846dfdf327530d1ea84210bff0c7/src/classlibnative/bcltype/bignum.h#L73 so it cannot be `const` or `constexpr`

mazong1123 replied:
It needs to be initialized in https://github.com/mazong1123/coreclr/blob/9b39ea6c0e28846dfdf327530d1ea84210bff0c7/src/classlibnative/bcltype/bignum.h#L73 so it probably cannot be `const` or `constexpr`

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,599 @@
+
+int BigNum::Compare(const BigNum& lhs, const BigNum& rhs)
+{
+    int lenDiff = (int)lhs.m_len - (int)rhs.m_len;
```
Could you leave a comment indicating that this is never expected to over/underflow due to the min/max of `m_len`?

Having an assert that `m_len` is never less than 0 or greater than `max length` would be good too

mazong1123 replied:
I think since `m_len` is UINT32, we just need to assert `max length`.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,599 @@
+        return lenDiff;
+    }
+
+    if (lhs.m_len == 0 && rhs.m_len == 0)
```
Just checking one of these values for `m_len == 0` should be sufficient (lenDiff is 0, so if one is zero, both must be). We can always have an assert to validate the other.

tannergooding commented on src/classlibnative/bcltype/bignum.cpp:
```diff
@@ -0,0 +1,599 @@
+    UINT32 lastIndex = lhs.m_len - 1;
+    INT32 currentIndex = lastIndex;
+
+    while (currentIndex >= 0)
```
`for (UINT32 currentIndex = lhs.m_len - 1; currentIndex >= 0; --currentIndex)` might be more readable?

tannergooding commented on src/classlibnative/bcltype/bignum.h:
```diff
@@ -0,0 +1,152 @@
+    static const UINT32 BIGSIZE = 35;
+    static const UINT32 UINT32POWER10NUM = 8;
+    static const UINT32 BIGPOWER10NUM = 6;
+    static constexpr UINT32 m_power10UInt32Table[UINT32POWER10NUM] = 
```
It doesn't need to be addressed now, but I think it would be worth investigating if holding up to 10^9 in here and then incrementing `m_power10BigNumTable` by 10, rather than 8 would be more efficient (I think it should result in fewer overall lookups).

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,288 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    // ========================================================================================================================================
+    // This implementation is based on the paper: https://www.cs.indiana.edu/~dyb/pubs/FP-Printing-PLDI96.pdf
+    // Besides the paper, some of the code and ideas are modified from http://www.ryanjuckett.com/programming/printing-floating-point-numbers/
+    // You must read these two marterials to fully understand the code.
```
`materials`

tannergooding commented on src/classlibnative/bcltype/number.cpp:
```diff
@@ -134,8 +135,288 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+    //
+    // 0.30102999566398119521373889472449 = log10V2
+    // 0.69 = 1 - log10V2 - epsilon (a small number account for drift of floating point multiplication)
+    int k = (int)(ceil(double((int)mantissaHighBitIdx + e) * 0.30102999566398119521373889472449 - 0.69));
```
nit, it would be more readable to declare `log10v2` and 0.69` as constants.

mazong1123 commented:
@tannergooding I've updated code according to the feedback. Please take a look. Thanks!

mazong1123 commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline
@dotnet-bot test Tizen armel Cross Release Build

mazong1123 commented:
> Do I need to put the license mentioned in #12894 (comment) anywhere?
At the beginning of number.cpp or DoubleToNumberWorker?
I am asking and will get back to you.

@tarekgh Do we have any update?

tarekgh commented:
> Do we have any update?

We are still following up here. looks there is some process we need to follow. we'll let you know what needs to be done when having more details. Thanks for your patience.

tarekgh commented:
We are still following up here. we'll update as soon as we have news.

tarekgh commented:
https://github.com/dotnet/coreclr/pull/13920

tarekgh commented:
@mazong1123 thanks a lot for your effort and getting this ready. finally, I am going to merge it :-)

jkotas commented:
This introduced build break on Windows ARM (this config does not build by default in CI):

`error C3861: 'BitScanReverse64': identifier not found [D:\j\workspace\arm_cross_che---9fe37e18\bin\obj\Windows_NT.arm.Checked\src\classlibnative\bcltype\bcltype.vcxproj]`

jkotas commented:
Fix: https://github.com/dotnet/coreclr/pull/13932

mazong1123 commented:
@jkotas how did you find the build break? Is there any way to check arm build in CI?

jkotas commented:
No worries. I happened to notice it on a different job. There are custom jobs that can be triggered manually based on the nature of the change. 

@dotnet-bot help 

prints the list. You will find the Windows ARM build somewhere in there. Obviously, it is not effective to run all of them for each job.
